README.md
# adventure_chess
simple cpp chess program


Na co dzień nie zajmuję się programowaniem w C++, więc aby nie zapomnieć tego 
języka wymyśliłem projekt, który ciągle mogę rozwijać, jeżeli tylko znajdę na to
ochotę i chwilę wolnego czasu. Wersja podstawowa, czyli Adventure Chess v.028, 
zajęła mi 28 dni pracy (stąd numeracja wersji) i została zakończona 20 listopada 
2008 roku. Nie była to ciągła praca, lecz tu i ówdzie wyrwane pół godziny, czy 
to w długie szwedzkie wieczory, czy w przewie meczu piłkarskiego, czy w 
deszczową sobotę.
Niektórzy mówią, że szachy są trudnym tematem do programowania, a jak się 
okazało temat można zamknąć już w niecałych 2000 lini kodu. Jednakże zdaję 
sobie sprawę, że liczba tych linii będzie ciągle rosła wraz ze stopniem 
zaawansowania projektu.
Wszystkich chętnych do pomocy, czy to przy pisaniu, czy to przy testowaniu 
zapraszam do współpracy.
Aktualności:
Aktualnie pracuję nad rozwojem silnika. W dalszej kolejności dostosuję program 
do środowiska graficznego Winboard/XBoard. W wersji Adventure Chess v.0.039 
zostały usunięte błędy zauważone podczas testowania Adventure Chess v.0.031. 


changelog
2008-12-06
Adventure Chess 0.039 delivered
Bugs fixed:
- Queen move from left edge to right edge of board.
- Check in three moves.
Features:
- A little engine tuning.

2008-11-23
Adventure Chess 0.031 delivered
Bugs fixed:
- Castling under check.
- Knight check from edges of the board.
- Wall checking during knight, rook, queen and king moves fixed
Features:
- Time of engine thinking measurement added.

2008-11-20
Adventure Chess 0.028 delivered - this is first "playable" version of the game


test.cc
/***************************************************************************
 *   Copyright (C) 2007 by raso                                            *
 *   Rafal Sowiak (sowiakr@interia.pl) - www.adventure.go.pl               *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <vector>

#include "engine.h"
//#include <teksty_en.h>//if language=english then include teksty_en.h otherwise include teksty_pl.h


using namespace std;

int main(int argc, char *argv[])
{
//declare
  int _moc=0;//help value
  int dalej=0;//checking end of program (0 - end of program)
  szachownica board(1);//object of szachownica with constructor
//program label
  cout << board.TITLE;
//board showing
  board.show();
//main loop
  do
  {
    if (board.black_or_white)
      { cout << board.WHITE_MOVE;
      }
    else
      { cout << board.BLACK_MOVE;
      };
    cin >> board.move;
//last statement rewriting
    if (board.DEBUG)
     cout << board.move << endl;
//end game checking
  dalej=board.move_ok(board.move,_moc);
//  cout << dalej << endl;
  if  (dalej==100)
       cout << board.BAD_MOVE << endl;
  else if  (dalej==101)
       cout << board.BAD_MOVE_CHECK << endl;
  else if (dalej!=3 && dalej)
       board.engine();
  }
  while (dalej);
//end of program
return 0;
}


engine.h
/***************************************************************************
 *   Copyright (C) 2007 by raso                                            *
 *   Rafal Sowiak (sowiakr@interia.pl) - www.adventure.go.pl               *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include "teksty_en.h"

void szachownica::engine()
{
  int MAXmoc=-1000;
  int MAXLOOP=2;
  int LOOP=0;
  int _moc;//help variable
  string MAXmove="----";
//single loop!!!!!!!!!!!!!!!!!!!!!!!!
  engine_loop(1,MAXmoc,MAXmove,LOOP,MAXLOOP);
//-----------------------------------------------------------
  if (DEBUG)
  {
    cout<<MAXmove<<endl;
    cout<<MAXmoc<<endl;
  }
//-----------------------------------------------------------
//show of engine move
  if (black_or_white)
    { cout <<"  "<< WHITE_MOVE;
    }
  else
    { cout <<"  "<< BLACK_MOVE;
    };
  cout<<MAXmove<<endl;
  if (MAXmove!="----")
    {
      move_ok(MAXmove,_moc);
      if (MAXmoc==1000)
         cout<<LOSE<<endl;
    }
  else
    { cout<<WIN<<endl;
    }
};

void szachownica::engine_loop(unsigned int _black_or_white,int &MAXmoc,string &MAXmove, int &LOOP, int MAXLOOP)
{
  int _i,_j,_k,_l;
  int TotalMoc,MAXmoc_white;
  string ActMove;
  unsigned int from,to,ffrom,tto;
  int moc;//power of move
//  int _param;
  string _fig1,_fig2,_fig3,_fig4,_move;
  int _wynik;
  if (_black_or_white)
    { _fig1="pqkrbn";
      _fig2="PQRBN.";
      _fig3="PQKRBN";
      _fig4="pqrbn.";
//      _param=0;
    }
  else
    { _fig1="PQKRBN";
      _fig2="pqrbn.";
      _fig3="pqkrbn";
      _fig4="PQRBN.";
//      _param=1;
    };
  LOOP++;
//-----------------------------------------------------------------------
  for (_i=0;_i<64;_i++)
  { if (_fig1.find(board_tab[_i])!=_fig1.npos)
      {  from=_i;
         for (_j=0;_j<64;_j++)
           { if (_fig2.find(board_tab[_j])!=_fig2.npos)
               { to=_j;
                 _move=move_tab[from];
                 _move+=move_tab[to];
                 moc=0;
                 _wynik=move_ok(_move,moc);
//--------------------------------------------------------------------
//making move
                 if (_wynik==1 || _wynik==2)
                  {
//power of machine move calculation
                    TotalMoc=moc;
                    ActMove=_move;
                    MAXmoc_white = 0;
//checking man move
                          _wynik=0;
//--------------------------------------------------------------------
                          for (_k=0;_k<64;_k++)
                          { if (_fig3.find(board_tab[_k])!=_fig3.npos)
                              {  ffrom=_k;
                                for (_l=0;_l<64;_l++)
                                  { if (_fig4.find(board_tab[_l])!=_fig4.npos)
                                      { tto=_l;
                                        _move=move_tab[ffrom];
                                        _move+=move_tab[tto];
                                        moc=0;
                                        _wynik=move_ok(_move,moc);
//making move
                                        if (_wynik==1 || _wynik==2)
                                          {
//power of man move calculation
                                              if (moc>MAXmoc_white)
                                              MAXmoc_white=moc;
//level of game choosing - future option
//                                              MAXmoc_white=0;
//------------------------------------------------------------------
                                            back(0);
                                          };
                                      };
                                  };
                              };
                          };

//white move end of calculation-------------------------------------------
// not aplicable now :)///////////////////////////////////////////////////
//                    if (LOOP<MAXLOOP)
//                        engine_loop(_param,MAXmoc,MAXmove,LOOP,MAXLOOP);
//////////////////////////////////////////////////////////////////////////
//machine wins
                    if (MAXmoc_white==0)
                      {
                        MAXmoc=1000;
                        MAXmove=ActMove;
                      }
//------------------------------------------------------------------------
                    else
                      {
                        TotalMoc-=MAXmoc_white;
                        if (TotalMoc>MAXmoc)
                          { MAXmoc=TotalMoc;
                            MAXmove=ActMove;
                          };
                      };
                    back(1);
                  };
               };
           };
      };
  };
  LOOP--;
//-------------------------------------------------------------
};

//***********************************************************************************************************
int szachownica::power_of_move(int to)
//***********************************************************************************************************
{
  int _wynik=0;
/*****************************************/
  int PowP=1;
  int PowB=6;
  int PowN=6;
  int PowR=11;
  int PowQ=16;
  int PowK=21;
/*****************************************/
//power of biting*************************/
  if (board_tab[to]=='p' || board_tab[to]=='P')
    _wynik+=PowP;
  else if (board_tab[to]=='b' || board_tab[to]=='B')
    _wynik+=PowB;
  else if (board_tab[to]=='n' || board_tab[to]=='N')
    _wynik+=PowN;
  else if (board_tab[to]=='r' || board_tab[to]=='R')
    _wynik+=PowR;
  else if (board_tab[to]=='q' || board_tab[to]=='Q')
    _wynik+=PowQ;
//power of board pole
/*
11111111
12222221
12333321
12345321
12345321
12333321
12222221
11111111
*/
  switch (to)
  {
    case 28:
    case 36:
     _wynik+=4;
     break;
    case 27:
    case 35:
     _wynik+=3;
     break;
    case 18:
    case 19:
    case 20:
    case 21:
    case 26:
    case 29:
    case 34:
    case 37:
    case 42:
    case 43:
    case 44:
    case 45:
    _wynik+=2;
     break;
    case 9:
    case 10:
    case 11:
    case 12:
    case 13:
    case 14:
    case 17:
    case 22:
    case 25:
    case 30:
    case 33:
    case 38:
    case 41:
    case 46:
    case 49:
    case 50:
    case 51:
    case 52:
    case 53:
    case 54:
    _wynik+=1;
     break;
  };
  return _wynik;
}


teksty_en.h
/***************************************************************************
 *   Copyright (C) 2007 by raso                                            *
 *   Rafal Sowiak (sowiakr@interia.pl) - www.adventure.go.pl               *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include "szach.h"

szachownica::szachownica(int _i)//constructor for szachownica class
{
//from 1 august 2007
TITLE="Adventure chess v.0,028b20081120 \nCopyleft Rafal Sowiak 2007-2008 (www.adventure.host22.com)\nType \"help\" for command list\n";//title statement
BUG_R="Any bugs please report to: rafalowapoczta@gmail.com";
NO_MOVIES="There were no movies yet.";
LIST="list";//printing list of moves
BACK="back";//printing list of moves
QUIT="quit";//quit of game statement
SHOW="show";//printout of board
SAVE="save";//saving game to file
LOAD="load";//loading game from file
HELP="help";//help screen
BAD_MOVE="Bad move!";//bad move printout
BAD_MOVE_CHECK="Bad move! Your king is checked.";//bad move printout
BAD_SAVE="Game not saved: error opening file";//faulty save
BAD_LOAD="Game not loaded: error opening file";//faulty save
WIN="You won! The game is finished.";//faulty load
LOSE="You lost! You are checkmated.";//faulty load
//FILENAME="Please enter the filename: ";
GOOD_SAVE="Game saved in file ";//proper save
GOOD_LOAD="Game loaded from file ";//proper load
WHITE_MOVE="White move: ";//white move label
BLACK_MOVE="Black move: ";//black move label
H_LABEL="Description of commands:";//help screen label
H_HELP="help - print this screen";
H_SHOW="show - print board actual arrangement";
H_LIST="list - print list of movies";
H_BACK="back - back last move";
H_SAVE="save <filename> - save game in file";
H_LOAD="load <filename> - load game from file";
H_QUIT="quit - end of this game end quit program";
FLIGHT_BIT=0;//set flight bit possibility to false
WCASTLING_W=0;
WCASTLING_B=0;
DEBUG=0;

black_or_white=_i;//set black_or_white move - future option
board_tab="rnbqkbnrpppppppp................................PPPPPPPPRNBQKBNR";//definition of board
/*
board_tab="rnbqkbnr
           pppppppp
           ........
           ........
           ........
           ........
           PPPPPPPP
           RNBQKBNR";//definition of board
*/

};
